### PR TITLE
feat: bin/cancel-all-pending-txs.js

### PR DESCRIPTION
**Pull request by bajtos**
WIP

### CODE REVIEW

**Comment by bajtos** [comment]
@juliangruber ready for your review. 

I managed to cancel the oldest TX.

The second run failed with "transaction execution reverted", but it still cancelled the transaction 🤔 

```
Error: transaction execution reverted (action="sendTransaction", data=null, reason=null, invocation=null, revert=null, transaction={ "data": "", "from": "0x4EcdC893Beb09121E4F5cBba469D33F5fF618442", "to": "0x4EcdC893Beb09121E4F5cBba469D33F5fF618442" }, receipt={ "_type": "TransactionReceipt", "blobGasPrice": null, "blobGasUsed": null, "blockHash": "0x71503ee0ecffdddfe45de3a4014ccd731774ec96390d50b7e8335e3c8b19802b", "blockNumber": 4284431, "contractAddress": null, "cumulativeGasUsed": "0", "from": "0x4EcdC893Beb09121E4F5cBba469D33F5fF618442", "gasPrice": "590849", "gasUsed": "1298840", "hash": "0x0be03a6d746686a0179a89dbcb051d166f25b62c27513af6ddc62feab9b8eaeb", "index": 71, "logs": [  ], "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000", "root": "0x0000000000000000000000000000000000000000000000000000000000000000", "status": 0, "to": "0x4EcdC893Beb09121E4F5cBba469D33F5fF618442" }, code=CALL_EXCEPTION, version=6.13.2)
```

The third run failed with the same error, but cancelled the transaction anyways.

It makes me wonder what to do about this. Should I not wait for TX confirmations and cancel all of them in one go? Or maybe in batches of 10 TX?

**Comment by bajtos** [comment]
Example of an accepted TX:
https://filfox.info/en/message/0x728719407de17212ea6f803d7d46cd0260e55d92b433d9901d1b6b04aa34ffb5?t=1

Example of rejected TX:
https://filfox.info/en/message/0x0be03a6d746686a0179a89dbcb051d166f25b62c27513af6ddc62feab9b8eaeb?t=1
> Status: SysErrOutOfGas

File: bin/cancel-all-pending-txs.js [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+  const nonce = tx.nonce
+
+  console.log(' - SENDING THE REPLACEMENT TRANSACTION')
+  try {
+    const replacementTx = await signer.sendTransaction({
+      to: signer.address,
+      value: 0,
+      nonce,
+      gasLimit: Math.ceil(gasUsed * 1.1),
+      maxFeePerGas: gasFeeCap,
```
Comment (juliangruber) [inline bin/cancel-all-pending-txs.js:208]: This patch is missing: https://github.com/filecoin-station/spark-evaluate/blob/e4358f466c53c5f457c4afcade70b3f47a95c128/lib/cancel-stuck-txs.js#L111-L112

**Review by juliangruber — CHANGES_REQUESTED** [review]

**Comment by juliangruber** [comment]
> I managed to cancel the oldest TX.
> [...]
> The second run failed with "transaction execution reverted", but it still cancelled the transaction 🤔
> [...]
> The third run failed with the same error, but cancelled the transaction anyways.
> 
> It makes me wonder what to do about this. Should I not wait for TX confirmations and cancel all of them in one go? Or maybe in batches of 10 TX?

If we don't wait for confirmations it will be hard to tell whether the script succeeded, and why. I think we need to see the errors in the logs, when there are some

**Comment by juliangruber** [comment]
Idk though why this happens

**Comment by juliangruber** [comment]
> An error we are not handling yet, I don't know what to do when this happens. Do you perhaps know, @juliangruber ?
> 
> ```
> failed to lookup Eth Txn 0x6533f4eb56d9857a1b28dfc99a645e8505987fb73ec6ee61bf2545b00f60ea48 as bafy2bzacec74uxibzyreox5uxlluzotk5fvwyn2evsvqavyugvkofjpkjinhi: 
> found message with equal nonce as the one we are looking for that is NOT a valid replacement message (
>   F:bafy2bzacec74uxibzyreox5uxlluzotk5fvwyn2evsvqavyugvkofjpkjinhi 
>   n 99056, 
>   TS: bafy2bzaceam7vdxb4oxa3vuhs4zicfiokuif6mxzdo6oukngah2ase6bkmj2c 
>   n99056
> )
> ```

I have seen this in logs before, but don't understand it yet either

**Review by juliangruber — APPROVED** [review]
Are you able to completely clear up the mempool with this script?

**Comment by bajtos** [comment]
Yes, the mempool was cleared.

File: bin/cancel-all-pending-txs.js (hunk 1 above)
Comment (bajtos) [reply bin/cancel-all-pending-txs.js:208]: I'll fix that in a follow-up PR.

Comment (bajtos) [reply bin/cancel-all-pending-txs.js:208]: https://github.com/filecoin-station/utils/pull/2

**Pull request merged**